pkg/cmd/experimental/router/router.go

```diff
@@ -0,0 +1,243 @@
+			}
+
+			ports, err := app.ContainerPortsFromString(cfg.Ports)
+			if err != nil {
+				glog.Fatal(err)
+			}
+
+			label := map[string]string{"router": name}
+			if cfg.Labels != defaultLabel {
+				valid, remove, err := app.LabelsFromSpec(strings.Split(cfg.Labels, ","))
+				if err != nil {
+					glog.Fatal(err)
+				}
+				if len(remove) > 0 {
+					glog.Fatalf("You may not pass negative labels in %q", cfg.Labels)
+				}
+				label = valid
+			}
+
+			image := cfg.ImageTemplate.ExpandOrDie(cfg.Type)
+
+			namespace, err := f.OpenShiftClientConfig.Namespace()
+			if err != nil {
+				glog.Fatalf("Error getting client: %v", err)
+			}
+			_, kclient, err := f.Clients(cmd)
+			if err != nil {
+				glog.Fatalf("Error getting client: %v", err)
+			}
+
+			p, output, err := cmdutil.PrinterForCommand(cmd)
+			if err != nil {
+				glog.Fatalf("Unable to configure printer: %v", err)
+			}
+
+			generate := output
+			if !generate {
+				_, err = kclient.Services(namespace).Get(name)
+				if err != nil {
+					if !errors.IsNotFound(err) {
+						glog.Fatalf("Can't check for existing router %q: %v", name, err)
+					}
+					generate = true
+				}
+			}
+
+			if generate {
+				// create new router
+
+				objects := []runtime.Object{
+					&dapi.DeploymentConfig{
+						ObjectMeta: kapi.ObjectMeta{
+							Name:   name,
+							Labels: label,
+						},
+						Triggers: []dapi.DeploymentTriggerPolicy{
+							{Type: dapi.DeploymentTriggerOnConfigChange},
+						},
+						Template: dapi.DeploymentTemplate{
+							Strategy: dapi.DeploymentStrategy{
+								Type: dapi.DeploymentStrategyTypeRecreate,
+							},
+							ControllerTemplate: kapi.ReplicationControllerSpec{
+								Replicas: cfg.Replicas,
+								Selector: label,
+								Template: &kapi.PodTemplateSpec{
+									ObjectMeta: kapi.ObjectMeta{Labels: label},
+									Spec: kapi.PodSpec{
+										Containers: []kapi.Container{
+											{
+												Name:  "router",
+												Image: image,
+												Ports: ports,
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				}
+				objects = app.AddServices(objects)
+				// TODO: label all created objects with the same label - router=<name>
+				list := &kapi.List{Items: objects}
+
+				if output {
+					if err := p.PrintObj(list, out); err != nil {
+						glog.Fatalf("Unable to print object: %v", err)
+					}
+					return
+				}
+
+				if !cfg.Create {
+					glog.Fatalf("Router %q does not exist (no service). Pass --create to install.", name)
+				}
+
+				bulk := configcmd.Bulk{
+					Factory: f.Factory,
+					Command: cmd,
+					After:   configcmd.NewPrintNameOrErrorAfter(out, os.Stderr),
+				}
+				if errs := bulk.Create(list, namespace); len(errs) != 0 {
+					os.Exit(1)
+				}
+				return
+			}
+
+			glog.Infof("Router %q service exists", name)
+		},
+	}
+
+	cmd.Flags().StringVar(&cfg.Type, "type", "haproxy-router", "The type of router to use - if you specify --images this flag may be ignored.")
```

Comment:
When you say "type" you probably mean "image name" right? Or at least partial image name...

Would like it to be clear that this is not just something from a list of tokens somewhere but actually feeds directly into which image is deployed in the container.